cluster/connector.go

```diff
@@ -0,0 +1,187 @@
+package cluster
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+
+	icid "github.com/ipfs/go-cid"
+	"github.com/ipfs/go-ipfs/core"
+	"github.com/ipfs/go-ipfs/core/coreapi"
+	"github.com/ipfs/go-ipfs/core/corerepo"
+	"github.com/ipfs/go-ipfs/pin"
+	iface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/ipfs/interface-go-ipfs-core/options"
+	"github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/ipfs/ipfs-cluster/api"
+	corepeer "github.com/libp2p/go-libp2p-core/peer"
+	rpc "github.com/libp2p/go-libp2p-gorpc"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type Connector struct {
+	node  *core.IpfsNode
+	api   iface.CoreAPI
+	peers func(ctx context.Context) []*api.ID
+}
+
+func NewConnector(node *core.IpfsNode, peers func(ctx context.Context) []*api.ID) (*Connector, error) {
+	capi, err := coreapi.NewCoreAPI(node)
+	if err != nil {
+		return nil, err
+	}
+	return &Connector{
+		node:  node,
+		api:   capi,
+		peers: peers,
+	}, nil
+}
+
+func (c *Connector) ID(context.Context) (*api.IPFSID, error) {
+	var addrs []api.Multiaddr
+	for _, addr := range c.node.PeerHost.Addrs() {
+		addrs = append(addrs, api.Multiaddr{Multiaddr: addr})
+	}
+	return &api.IPFSID{
+		ID:        c.node.Identity,
+		Addresses: addrs,
+	}, nil
+}
+
+func (c *Connector) SetClient(client *rpc.Client) {
+	// noop
+}
+
+func (c *Connector) Shutdown(ctx context.Context) error {
+	// noop
+	return nil
+}
+
+// @todo handle maxDepth
+func (c *Connector) Pin(ctx context.Context, cid icid.Cid, maxDepth int) error {
+	return c.api.Pin().Add(ctx, path.New(cid.String()))
+}
+
+func (c *Connector) Unpin(ctx context.Context, cid icid.Cid) error {
+	return c.api.Pin().Rm(ctx, path.New(cid.String()))
+}
+
+func (c *Connector) PinLsCid(ctx context.Context, cid icid.Cid) (api.IPFSPinStatus, error) {
+	pins, err := c.node.Pinning.CheckIfPinned(cid)
+	if err != nil {
+		return api.IPFSPinStatusError, err
+	}
+	if len(pins) == 0 {
+		return api.IPFSPinStatusError, fmt.Errorf("invalid pin check result")
+	}
+	return c.pinModeToStatus(pins[0].Mode), nil
+}
+
+func (c *Connector) PinLs(ctx context.Context, typeFilter string) (map[string]api.IPFSPinStatus, error) {
+	pins, err := c.api.Pin().Ls(ctx, c.pinFilterToOption(typeFilter))
+	if err != nil {
+		return nil, err
+	}
+	statusMap := make(map[string]api.IPFSPinStatus)
+	for _, p := range pins {
+		mode, ok := pin.StringToMode(p.Type())
+		if !ok {
+			continue
+		}
+		statusMap[p.Path().String()] = c.pinModeToStatus(mode)
+	}
+	return statusMap, nil
+}
+
+func (c *Connector) ConnectSwarms(ctx context.Context) error {
```

Comment:
For information: this will get triggered when a new cafe "bootstraps" (uses `Cluster.Join(<peer maddr>)`). It will be called in the cluster peer it bootstraps to. If textile never calls `Join()` to introduce new peers to the Cluster*, then it can be a noop.

*`Join()` is not strictly necessary with CRDTs, but it is a way of getting two libp2p hosts in the same cluster connected (and this should 1) allow pubsub to start working (otherwise the peer may never receive pubsub messages because it doesn't know any of the other peers subscribed to them), 2) Potentially allow dht-discovery of other cluster peers and better connectivity.